### ISSUE TITLE
Emoji render no color

### ISSUE THREAD

**Issue by letungcntt** [flutter/flutter#100964]
On desktop, sometimes emojis render no color. I can check it, with macOS or windows. On windows, the problem is more serious

https://user-images.githubusercontent.com/46608407/160598573-7816262a-98c3-4fcd-baf7-ab272f24678e.mov

**Comment by daadu** [comment]
I guess, these are rendered based on Font.

**Comment by letungcntt** [comment]
I hope it's the answer. but sometimes emojis render right, sometimes render wrong.

https://user-images.githubusercontent.com/46608407/160612596-158e1a23-7953-4f51-bfec-d24d651688d4.mov

**Comment by letungcntt** [comment]
On Windows, emojis can't render flags color: 
![image](https://user-images.githubusercontent.com/46608407/160623276-002681ba-cb39-486c-8769-b56548294de3.png)

**Comment by letungcntt** [comment]
@daadu Thanks for the answer, I have pulled and added this font https://github.com/googlefonts/noto-emoji/tree/main/fonts.
By macOS, the problem doesn't fix. By Windows, most emojis had rendered color. However, emojis in Flag, can't render color

**Comment by daadu** [comment]
Did you tired windows compatible version?

**Comment by letungcntt** [comment]
yeah. I had used `NotoColorEmoji_WindowsCompatible.ttf` on windows. Normally, emojis cant render color but after I added `NotoColorEmoji_WindowsCompatible.ttf` most of the emojis rendered color, except emojis in tag flags

**Comment by maheshmnj** [comment]
cc: @LongCatIsLooong

**Comment by letungcntt** [comment]
On Windows. i hope emojis flags can render color

**Comment by letungcntt** [comment]
@maheshmnj The problem had existed in flutter 3.0 stable. Has this problem been solved in master?

**Comment by letungcntt** [comment]
@LongCatIsLooong Any update this problems?

**Comment by jason-simmons** [comment]
See https://github.com/flutter/flutter/issues/107538#issuecomment-1190760461

If the fonts listed in the `TextStyle` do not provide a character, then Flutter needs to search the host system for a font that does.  If multiple fonts provide that character, then the selected font may not be the one that you prefer.

On my macOS setup, both Apple Color Emoji and Zapf Dingbats provide the "red heart emoji" character.  If that character is the first character in the string, then Skia's font manager implementation for macOS is selecting Zapf Dingbats, which renders it as a black glyph.

If the first character is the yellow heart, then Skia selects Apple Color Emoji, which is the only available font that provides that character.  The text renderer can then continue using Apple Color Emoji to render the red heart character.

Searching for host system fonts may not behave consistently across different host platforms.  If the app wants to control how these characters look, then it should embed fonts that render the characters as intended.

**Comment by letungcntt** [comment]
> See [#107538 (comment)](https://github.com/flutter/flutter/issues/107538#issuecomment-1190760461)
> 
> If the fonts listed in the `TextStyle` do not provide a character, then Flutter needs to search the host system for a font that does. If multiple fonts provide that character, then the selected font may not be the one that you prefer.
> 
> On my macOS setup, both Apple Color Emoji and Zapf Dingbats provide the "red heart emoji" character. If that character is the first character in the string, then Skia's font manager implementation for macOS is selecting Zapf Dingbats, which renders it as a black glyph.
> 
> If the first character is the yellow heart, then Skia selects Apple Color Emoji, which is the only available font that provides that character. The text renderer can then continue using Apple Color Emoji to render the red heart character.
> 
> Searching for host system fonts may not behave consistently across different host platforms. If the app wants to control how these characters look, then it should embed fonts that render the characters as intended.

Thank you, but how to set default font my_app to Apple Color Emoji. When i log this font default, it print `.AppleColorEmoji`. I has import assets "NotoColorEmoji" and use with TextStyle but the problem not fixed.

**Comment by nnhoang215** [comment]
@jason-simmons Yes, so @letungcntt and I did manage to overcome this by setting the fontFamilyFallback value to ["Roboto", "Apple Color Emoji"]. This worked because these two fonts do not overlap. However, we made changes to the framework in text_style.dart to apply such changes to the whole application. Is there a better way to do this without messing with the framework?

**Comment by 12people** [comment]
@daadu

> Try using https://github.com/googlefonts/noto-emoji/tree/main/fonts -as font for colored emoji.

Unfortunately, Google's Noto fonts don't support the SBIX format that macOS and iOS use, so it won't work for these platforms. See https://github.com/googlefonts/noto-emoji/issues/72 for more info.

**Comment by gspencergoog** [comment]
@jason-simmons Would it be appropriate to forward this issue to the Skia team to see if there is anything that can be improved in the font fallback/selection process? If so, could you reach out to them?

(Desktop Triage)

**Comment by knopp** [comment]
There is another issue here, not just fallbacks. Skia doesn't seem to observe the emoji variation selector (U+FE0F). For example the red heart emoji (U+2665) is often present in fonts as black heart. So to make it an emoji the U+FE0F selector is appended to it. Example: ♥♥️

Should render as:

<img width="40" alt="Screenshot 2023-02-24 at 11 50 30" src="https://user-images.githubusercontent.com/96958/221160764-1cbb5db9-1768-4415-9867-4dd7119c16bf.png">

**Comment by knopp** [comment]
From a brief look into shaping in skia it seems that `FontMgrRunIterator` is only looking at the glyph itself when determining the font, ignoring that the next glyph might specify the variation. Also the `SkFontMgr::matchFamilyStyleCharacter` used to determine the font doesn't have any way to take the variant into account.

I think this is the [PR](https://hg.mozilla.org/mozilla-central/rev/65421ad99b13) that addressed this issue in Firefox.  When getting font for character next character is passed and if the next character is emoji variant selector the platform emoji font is used first.

**Comment by knopp** [comment]
Small correction. `FontMgrRunIterator` does not seem to be used by `OneLineShaper`. `OneLineShaper` is responsible for getting the fallback font from `FrontCollection`.

The skia bug regarding U+FE0F is tracked here: https://bugs.chromium.org/p/skia/issues/detail?id=11755

**Comment by ohir** [comment]
*A note to whomever will look into it deeper: Esp. from the skia team.*

Apple's emoji picker produces many *unqualified* by UCD two codepoints sequences, where standard says 'FE0F' is mandated, and Apple's picker produces single emoji codepoint.
Examples:
```
invalid sequence |⏸| 23F8. Should be 23F8 FE0F. Omitted!
invalid sequence |⏹| 23F9. Should be 23F9 FE0F. Omitted!
invalid sequence |⏺| 23FA. Should be 23FA FE0F. Omitted!
invalid sequence |⏭| 23ED. Should be 23ED FE0F. Omitted!
invalid sequence |⏮| 23EE. Should be 23EE FE0F. Omitted!
```
(also recently macOS/iOS picker allows user to produce incompatible two-person ZWJ sequences,
but that's not relevant here).
Cc: [skia bugtracker](https://bugs.chromium.org/p/skia/issues/detail?id=11755)

**Comment by knopp** [comment]
> _A note to whomever will look into it deeper: Esp. from the skia team._
> 
> Apple's emoji picker produces many _unqualified_ by UCD two codepoints sequences, where standard says 'FE0F' is mandated, and Apple's picker produces single emoji codepoint. Examples:
> 
> ```
> invalid sequence |⏸| 23F8. Should be 23F8 FE0F. Omitted!
> invalid sequence |⏹| 23F9. Should be 23F9 FE0F. Omitted!
> invalid sequence |⏺| 23FA. Should be 23FA FE0F. Omitted!
> invalid sequence |⏭| 23ED. Should be 23ED FE0F. Omitted!
> invalid sequence |⏮| 23EE. Should be 23EE FE0F. Omitted!
> ```
> 
> (also recently macOS/iOS picker allows user to produce incompatible two-person ZWJ sequences, but that's not relevant here). Cc: [skia bugtracker](https://bugs.chromium.org/p/skia/issues/detail?id=11755)

This is not quite my experience. For example for the "stop" the emoji picker can provide both ⏹️ (which as the FE0F selector) and ⏹ which doesn't. But for fast forward (⏩) it does omits the selector.

**Comment by EddieMataEwy** [comment]
For people on Windows, setting familyFontFallback didn't work for me.
What worked was setting app theme familyfont (you can also only set it for specific elements) to "Segoe UI Emoji". Normal text will still be rendered natively but emojis will now be rendered correctly.
Flutter falls back to "Segoe UI Variant" for anything that is not an emoji I believe.

**Comment by gnprice** [comment]
A previous thread on the same issue is #49627. A comment there (https://github.com/flutter/flutter/issues/49627#issuecomment-1219807409) by @JoshJuncker has a nice demonstration of how the order of emoji in a paragraph affects the issue:
![image](https://github.com/flutter/flutter/assets/28173/ce3891c0-cca2-4a53-ae3d-de7956490cda)

(I think https://github.com/flutter/flutter/issues/100964#issuecomment-1190778513 above explains why the order makes a difference.)

That comment also reports that the issue is found on Android, in addition to the other platforms already mentioned in this thread.

**Comment by MichalNemec** [comment]
macOS 14.6.1 (23G93)
flutter 
```
Flutter 3.24.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 603104015d (7 weeks ago) • 2024-10-24 08:01:25 -0700
Engine • revision db49896cf2
Tools • Dart 3.5.4 • DevTools 2.37.3
```
by using
```dart
ThemeData(
    brightness: brightness,
    fontFamilyFallback: ['Apple Color Emoji'],
    useMaterial3: true,
  )
```
in TextFormField i see black "red" heart.
in Text i see black "red" heart.
codeUnits are [9829, 65039]
<img width="99" alt="image" src="https://github.com/user-attachments/assets/5b7c8d75-b0ea-4aa4-8a68-f0b9c458ff1a">

<img width="55" alt="image" src="https://github.com/user-attachments/assets/3f1cca2d-2a1e-49f2-9395-461d800dd81c">

If i use fontFamily: 'Apple Color Emoji' then it works, but everything has super word spacing.
It seems like fontFamilyFallback does not work at all?
